GVFS/GVFS.Service/GvfsService.cs

```diff
@@ -408,8 +432,73 @@ private void CreateAndConfigureUpgradeLogDirectory()
 
             Directory.CreateDirectory(upgradeLogsPath, upgradeLogsSecurity);
 
-            // Ensure the ACLs are set correct on any files or directories that were already created (e.g. after upgrading VFS4G)
-            Directory.SetAccessControl(upgradeLogsPath, upgradeLogsSecurity);
+            try
+            {
+                // Ensure the ACLs are set correct on any files or directories that were already created (e.g. after upgrading VFS4G)
+                Directory.SetAccessControl(upgradeLogsPath, upgradeLogsSecurity);
+            }
+            catch (UnauthorizedAccessException e)
+            {
+                // This can happen when the Logs directory was created by a non-elevated user running 'gvfs upgrade'
+
+                EventMetadata metadata = new EventMetadata();
+                metadata.Add("Exception", e.ToString());
+                metadata.Add(
+                    TracingConstants.MessageKey.InfoMessage,
+                    $"{nameof(this.CreateAndConfigureUpgradeLogDirectory)}: UnauthorizedAccessException when setting log dir ACLs");
+                this.tracer.RelatedEvent(EventLevel.Informational, "LogDirACL_UnauthorizedAccessException", metadata);
+
+                this.MigrateUpgradeLogsToDirectoryWithFreshACLs();
+            }
+        }
+
+        private void MigrateUpgradeLogsToDirectoryWithFreshACLs()
+        {
+            string upgradeLogsPath = ProductUpgraderInfo.GetLogDirectoryPath();
+            string tempUpgradeLogsPath = Path.Combine(
+                Path.GetDirectoryName(upgradeLogsPath),
+                ProductUpgraderInfo.LogDirectory + "_" + Guid.NewGuid().ToString("N"));
+
+            this.tracer.RelatedInfo($"Renaming '{upgradeLogsPath}' to '{tempUpgradeLogsPath}'");
+            Directory.Move(upgradeLogsPath, tempUpgradeLogsPath);
+
+            this.tracer.RelatedInfo($"Creating new '{upgradeLogsPath}' directory with appropriate ACLs");
+            DirectorySecurity upgradeLogsSecurity = new DirectorySecurity();
+
+            // Protect the access rules from inheritance and remove any inherited rules
+            // (any manually added ACLs are left in place)
+            upgradeLogsSecurity.SetAccessRuleProtection(isProtected: true, preserveInheritance: false);
+
+            // Add ACLs for users and admins
+            WindowsFileSystem.AddUsersAccessRulesToDirectorySecurity(upgradeLogsSecurity, grantUsersModifyPermissions: true);

```

Comment:
There is some (non-obvious?) logic here that is pretty specific to adding the access rules to the Logs directory, but it is duplicated. Would there be value in pulling out the common logic so we do not need to ensure that we are adding the ACLs for the upgrade logs directory in 2 places?